loc_books/metadata.py

```diff
@@ -0,0 +1,263 @@
+import json
+import os
+from concurrent.futures import ThreadPoolExecutor as PoolExecutor
+from pathlib import Path
+from time import time
+
+import click
+import pandas as pd
+import requests
+from dateutil.parser import parse
+from furl import furl
+from pyrate_limiter import Duration, Limiter, RequestRate
+from requests_ratelimiter import LimiterSession
+from titlecase import titlecase
+from tqdm import tqdm
+
+from licensed_pile import logs
+
+data_path = Path(__file__).resolve().parent / "data"
+metadata_downloads_path = data_path / "downloads/metadata"
+metadata_exports_path = data_path / "exports/metadata"
+
+metadata_downloads_path.mkdir(parents=True, exist_ok=True)
+metadata_exports_path.mkdir(parents=True, exist_ok=True)
+
+filename_digits = 4
+
+logger = logs.get_logger("loc_books")
+
+
+class LocBooksMetadataDownloader:
+    def __init__(self, base_url, snapshot):
+        self.base_url = base_url
+        self.snapshot = snapshot
+        self.download_path = metadata_downloads_path / snapshot
+
+        self.burst_rate = RequestRate(10, Duration.SECOND * 10)
+        self.crawl_rate = RequestRate(60, Duration.MINUTE)
+        self.limiter = Limiter(self.burst_rate, self.crawl_rate)
+        self.session = LimiterSession(limiter=self.limiter)
+
+        self.date_facets = []
+        self.items_per_page = 100
+        self.progress_bar = tqdm(
+            total=0, position=0, delay=1, desc="Downloading metadata"
+        )
+        self.existing_pages_count = 0
+
+    def start_download(self):
+        base_furl = furl(self.base_url)
+        base_furl.args["fo"] = "json"
+        logger.info(f"Downloading metadata from URL: {base_furl.url}")
+        try:
+            response = self.session.get(base_furl.url)
+            response.raise_for_status()
+            json_data = response.json()
+            self.parse_date_facets(json_data)
+            for date_facet in self.date_facets:
+                self.download_date_facet(date_facet)
+        except requests.exceptions.RequestException as e:
+            logger.error("Failed to download base URL:", str(e))
+
+    def download_date_facet(self, date_facet):
+        facet_path = self.download_path / date_facet["year_range"]
+        facet_folder = facet_path.absolute()
+        facet_url = date_facet["link"]
+
+        os.makedirs(facet_folder, exist_ok=True)
+
+        self.download_page(facet_url, facet_folder, 1)
+        with open(
+            os.path.join(facet_folder, f"{str(1).zfill(filename_digits)}.json")
+        ) as f:
+            file_data = json.load(f)
+            total_pages = file_data["pagination"]["total"]
+
+        pages_to_download = self.check_existing_files(total_pages, facet_folder)
+        self.existing_pages_count += total_pages - len(pages_to_download)
+        self.progress_bar.total += len(pages_to_download)
+
+        with PoolExecutor(max_workers=10) as executor:
```

Comment:
Is there a specific reason to use this over `multiprocessing.dummy.Pool` like you did before?